subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/service/BuildCacheServicesConfiguration.java

```diff
@@ -17,19 +17,24 @@
 package org.gradle.caching.internal.controller.service;
 
 import org.gradle.caching.BuildCacheService;
+import org.gradle.caching.local.internal.LocalBuildCacheService;
+
+import javax.annotation.Nullable;
 
 public final class BuildCacheServicesConfiguration {
 
-    public final BuildCacheService local;
+    @Nullable
+    public final LocalBuildCacheService local;
     public final boolean localPush;
 
+    @Nullable
```

Comment:
We also don't access fields directly, except here. Since we do, we need these annotations. :/ I can change this to use getters. WDYT?

I'm not sure why IntelliJ doesn't flag the field assignment here as problematic, though:

```java
public final class BuildCacheServicesConfiguration {

    public final BuildCacheService remote;

    public BuildCacheServicesConfiguration(
        @Nullable BuildCacheService remote
    ) {
        this.remote = remote;
    }
}
```

Reply:
I think IDEA infers the nullability of fields automatically. I don't get any warnings when checking the fields for null without the annotations. That is why I would remove the annotation.

Reply:
Nor does the `remote` field inherit the nullability of the `remote` parameter without declaring the field itself as `@Nullable`.

Reply:
Here's what I did. I removed `@Nullable` here from the `remote` field, then went over to `DefaultBuildCacheController` and removed `@Nullable` from the the declaration of `toRemoteHandle()`'s `service` parameter. Now, with this if IntelliJ properly detected the nullability of the field here, then it should signal a problem with this call in the `DefaultBuildCacheController` constructor, complaining that `config.remote` can be null:

```java
        this.remote = toRemoteHandle(config.remote, config.remotePush, buildOperationExecutor, logStackTraces);
```

...but it doesn't until you add back the `@Nullable` annotation on `BuildCacheServicesConfiguration.remote`.

Reply:
I am getting a warning on the `@Nullable` annotation on the field: 
![image](https://user-images.githubusercontent.com/423186/61809354-66a77780-ae3d-11e9-974c-b090b71c0743.png)

Reply:
That's another issue. Anyway, I've refactored to use methods. That's what this code should have been doing since the beginning.

Notice though that we need annotations on the getters just as we need them on the fields. The only difference is that IntelliJ can correctly signal nullability problem with the getter:

![image](https://user-images.githubusercontent.com/495366/61810245-1d582780-ae3f-11e9-8966-1f3299bfff3b.png)